### PR TITLE
Refactor Widget Base Class

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <gordon.smith@lexisnexis.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
@mlzummo @jbrundage @dtsnell4 Please review.  Just a refactor of the Widget base class to break up the different functional pieces into their own source files.
http://rawgit.com/GordonSmith/Visualization/PLANE_REFACTOR/test/runner_rawgit.html

**Comment by mlzummo** [comment]
this is what we needed ... Axis can now extend PropertyExt

File: src/common/PropertyExt.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                    if (!(_ instanceof Object)) {
+                        console.log("Invalid value for '" + id + "':  " + _ + " expected " + type);
+                    }
+                    return _;
+                };
+                break;
+            case "widget":
+            case "widgetArray":
+                this.checkedAssign = function (_) {
+                    //  TODO  ---
```
Comment (mlzummo) [inline src/common/PropertyExt.js]: ???

Comment (mlzummo) [inline src/common/PropertyExt.js]: can we add one for propertyExt?

Comment (GordonSmith) [inline src/common/PropertyExt.js]: The TODO is for me to add a test that _ === WidgetArray (I don't have a clean way yet)

**Comment by mlzummo** [comment]
@GordonSmith can we add

**Comment by GordonSmith** [comment]
@mlzummo Can you elaborate, I am not sure what your thinking of?

**Comment by mlzummo** [comment]
@GordonSmith for the Axis obj that we are working on instead of using a whole widget .. with this refactor we would be able to just extends propertyExt class wouldnt we? and if thats the case then widgetarray wouldnt accept a propertyExy Obj we would need a new one and then be able to serialize that ...

File: src/common/PropertyExt.js (hunk 1 above)
Comment (mlzummo) [inline src/common/PropertyExt.js]: check that its a widget and that its an Instance of Widget?

**Comment by mlzummo** [comment]
@GordonSmith ... when will this be finished and pulled?

**Comment by GordonSmith** [comment]
Its waiting on a final review.

File: src/common/PropertyExt.js (hunk 1 above)
Comment (GordonSmith) [inline src/common/PropertyExt.js]: If I do that it will be in a __hpcc_debug section...

**Comment by GordonSmith** [comment]
@mlzummo Added type checking for widget + WidgetArray (can't use instanceof due to circular references...)

**Comment by mlzummo** [comment]
looks good

**Pull request merged**